build/chrony/patches/failfast.patch

```diff
@@ -0,0 +1,101 @@
+From 83f268396a8042f18d2524ae2e7bef8437061cc2 Mon Sep 17 00:00:00 2001
+From: Andy Fiddaman <illumos@fiddaman.net>
+Date: Thu, 27 Feb 2025 21:05:26 +0000
+Subject: [PATCH] Add failfast source option for faster fallback to local
+ source.
+
+This adds a new "failfast" option to the source definition commands
+(peer/server) that causes them to be marked as unselectable as soon
+as they are deemed unreachable (the last X attempts to communicate
+have all failed). This causes a rapid fallback to any configured
+local source.
+---
+ cmdparse.c  |  2 ++
+ reference.c |  6 ++++++
+ reference.h |  1 +
+ sources.c   | 11 +++++++++++
+ srcparams.h |  1 +
+ 5 files changed, 21 insertions(+)
+
+diff --git a/cmdparse.c b/cmdparse.c
+index 77447dc2..32742595 100644
+--- a/cmdparse.c
++++ b/cmdparse.c
+@@ -210,6 +210,8 @@ CPS_GetSelectOption(char *option)
+     return SRC_SELECT_REQUIRE;
+   } else if (!strcasecmp(option, "trust")) {
+     return SRC_SELECT_TRUST;
++  } else if (!strcasecmp(option, "failfast")) {
++    return SRC_SELECT_FAILFAST;
+   }
+   return 0;
+ }
+diff --git a/reference.c b/reference.c
+index efce3679..9da8963b 100644
+--- a/reference.c
++++ b/reference.c
+@@ -1285,6 +1285,12 @@ REF_DisableLocal(void)
+   LOG(LOGS_INFO, "%s local reference mode", "Disabled");
+ }
+ 
++int
++REF_LocalEnabled(void)
++{
++  return (enable_local_stratum);
++}
++
+ /* ================================================== */
+ 
+ #define LEAP_SECOND_CLOSE 5
+diff --git a/reference.h b/reference.h
+index 2eddcae9..ca30ea4a 100644
+--- a/reference.h
++++ b/reference.h
+@@ -187,6 +187,7 @@ extern void REF_ModifyMakestep(int limit, double threshold);
+ 
+ extern void REF_EnableLocal(int stratum, double distance, int orphan, double activate);
+ extern void REF_DisableLocal(void);
++extern int REF_LocalEnabled(void);
+ 
+ /* Check if either of the current raw and cooked time, and optionally a
+    provided timestamp with an offset, is close to a leap second */
+diff --git a/sources.c b/sources.c
+index e292e4d6..f6de91c7 100644
+--- a/sources.c
++++ b/sources.c
+@@ -960,6 +960,15 @@ SRC_SelectSource(SRC_Instance updated_inst)
+       continue;
+     }
+ 
++    /* If the source is configured to fail fast and is currently unreachable,
++     * mark it as unselectable. */
++    if ((sources[i]->sel_options & SRC_SELECT_FAILFAST) &&
++	sources[i]->reachability == 0 &&
++	sources[i]->reachability_size == SOURCE_REACH_BITS) {
```

Comment:
Are we checking against reachability here because this is indicating that we've basically increased the maximum amount of reachability information and that we haven't decided it's unreachable until basically all attempts have been made?

Reply:
Yes. `reachability` is a bitmask in which the lowest `reachability_size` bits are relevant so we only want to mark a fastfail source as unreachable when we've collected `SOURCE_REACH_BITS` (8) samples rather than when we first start polling a source.